In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%config InlineBackend.figure_format = 'retina'

## Q1

In [2]:
n_trials = nrows = 10_000
n_dice = ncols = 1

roll1 = np.random.choice([1, 2, 3, 4, 5, 6], n_trials * n_dice).reshape(nrows, ncols)
roll2 = np.random.choice([1, 2, 3, 4, 5, 6], n_trials * n_dice).reshape(nrows, ncols)

roll1
roll2

array([[6],
       [5],
       [2],
       ...,
       [1],
       [6],
       [3]])

In [3]:
doubles = roll1 == roll2
doubles

array([[ True],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [4]:
total_doubles = doubles.sum()
total_doubles

1673

In [5]:
doubles_rate = doubles.astype(int).mean()
doubles_rate

0.1673

The probability of rolling doubles when rolling two dice is .1609 out of 1, or ~16% of 100%

## Q2

In [6]:
n_trials = nrows = 10_000
n_coins = ncols = 8

flips = np.random.choice([0,1], n_trials * n_coins).reshape(nrows, ncols)
flips

array([[0, 1, 0, ..., 1, 1, 0],
       [0, 0, 0, ..., 1, 1, 0],
       [1, 1, 1, ..., 0, 1, 1],
       ...,
       [1, 0, 1, ..., 0, 1, 1],
       [0, 1, 1, ..., 0, 1, 0],
       [1, 1, 1, ..., 1, 1, 1]])

In [7]:
headcount = flips.sum(axis=1)
headcount

array([5, 3, 5, ..., 6, 3, 7])

In [8]:
success1 = headcount == 3
success1

array([False,  True, False, ..., False,  True, False])

In [9]:
success1.sum()

2216

In [10]:
success1_rate = success1.astype(int).mean()
success1_rate

0.2216

The probability of getting exactly 3 heads is ~22% of 100%

In [11]:
success2 = headcount > 3
success2

array([ True, False,  True, ...,  True, False,  True])

In [12]:
success2.sum()

6411

In [13]:
success2_rate = success2.astype(int).mean()
success2_rate

0.6411

The probability of getting more than 3 heads is ~64% of 100%

## Q3

In [14]:
n_trials = nrows = 10_000
n_alum = ncols = 2

alum = np.random.choice([1,2,3,4], n_trials * n_alum).reshape(nrows, ncols)
alum

array([[4, 2],
       [2, 3],
       [1, 4],
       ...,
       [3, 1],
       [2, 3],
       [3, 1]])

Let's say that a value of 1 means that the chosen student is from data science.

In [15]:
alum_select = alum.sum(axis=1)
alum_select

array([6, 5, 5, ..., 4, 5, 4])

In [16]:
double_ds = alum_select == 2
double_ds

array([False, False, False, ..., False, False, False])

In [17]:
double_ds.sum()

628

In [18]:
double_ds_rate = double_ds.astype(int).mean()
double_ds_rate

0.0628

There is a 6.24% probability of DS alumni being on two billboards passed in a row.

## Q4

In [19]:
n_trials = nrows = 10_000
n_pops = ncols = 5

pops_weekly=np.random.normal(3, 1.5, 50000).reshape(nrows,ncols)
pops_weekly

array([[ 3.12564128,  0.52405106,  4.22823753,  3.42991578,  2.93276044],
       [ 1.86971708,  4.94819201,  0.26645649,  0.84885894,  1.48605022],
       [ 2.87265152,  2.39078447,  3.37391803,  3.73575403,  3.47459535],
       ...,
       [ 5.33338777,  3.93392579,  3.49340665,  3.53959916,  0.99798227],
       [ 2.55267996,  2.02595929,  4.4982688 ,  2.47454424,  2.22820617],
       [ 2.71946556, -0.03834588,  1.59734957,  3.73992805,  0.78795418]])

In [20]:
friday=17-pops_weekly.sum(axis=1)
friday

array([ 2.7593939 ,  7.58072525,  1.15229661, ..., -0.29830164,
        3.22034154,  8.19364852])

In [21]:
pop_fri = friday > 0
pop_fri

array([ True,  True,  True, ..., False,  True,  True])

In [22]:
pop_fri_rate = pop_fri.astype(int).mean()
pop_fri_rate

0.7237

There is a ~72% probability that there will be poptarts available by Friday afternoon.

## Q5

In [23]:
n_trials = nrows = 10_000
n_people = ncols = 2

men = np.random.normal(178, 8, nrows)
women = np.random.normal(170, 6, nrows)
#men_and_women = [men,women]
#men_and_women

In [24]:
women_taller = men < women
women_taller

array([False, False,  True, ..., False, False,  True])

In [25]:
women_taller.astype(int).mean()

0.2189

There is ~21% probability of the woman being taller than the man.

## Q6

In [26]:
n_trials = nrows = 50
n_students = ncols = 1

installs = np.random.uniform(1,250, nrows).astype(int).reshape(nrows,ncols)
#installs

In [27]:
corrupt = installs == 1
#corrupt

In [28]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.0

In [29]:
(corrupt.sum(axis=1) == 0).mean()

1.0

Out of 50 students, ~2% (1 student) had a corrupted download.

About 98% chance of installation resulting in no corruption.

In [30]:
n_trials = nrows = 100
n_students = ncols = 1

installs = np.random.uniform(1,250, nrows).astype(int).reshape(nrows,ncols)
#installs

In [31]:
corrupt = installs == 1
#corrupt

In [32]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.0

In [33]:
(corrupt.sum(axis=1) == 0).mean()

1.0

~98% chance of no corruption

In [34]:
n_trials = nrows = 150
n_students = ncols = 1

installs = np.random.uniform(1,250, nrows).astype(int).reshape(nrows,ncols)
#installs

In [35]:
corrupt = installs == 1

In [36]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.006666666666666667

In [37]:
(corrupt.sum(axis=1) == 0).mean()

0.9933333333333333

~98% chance of no corruption

In [38]:
n_trials = nrows = 450
n_students = ncols = 1

installs = np.random.uniform(1,250, nrows).astype(int).reshape(nrows,ncols)
#installs

In [39]:
corrupt = installs == 1

In [40]:
corrupt_rate = corrupt.astype(int).mean()
corrupt_rate

0.006666666666666667

In [41]:
(corrupt.sum(axis=1) == 0).mean()

0.9933333333333333

~98% chance of no corruption

## Q7

In [42]:
n_trials = nrows = 10000
n_week = ncols = 7

truck_chance = np.random.random(nrows*ncols).reshape(nrows,ncols)
#truck_chance

In [43]:
truck_present = truck_chance <= .70
truck_present

array([[ True,  True, False, ..., False,  True,  True],
       [False,  True,  True, ...,  True,  True,  True],
       [ True,  True,  True, ..., False,  True,  True],
       ...,
       [False,  True,  True, ..., False,  True, False],
       [ True,  True,  True, ...,  True,  True,  True],
       [False, False,  True, ...,  True,  True,  True]])

In [44]:
truck_present_rate = truck_present.astype(int).mean()
truck_present_rate

0.7002

In [45]:
(truck_present.sum(axis=1) <= 4).mean()

0.3535

There is a 35% probability that a truck is present 4 or less days of the week.

In [46]:
(truck_present.sum(axis=1) >=1).mean()

0.9995

There is a 99.95% probability that a truck will be present this week.

## Q8

In [56]:
n_trials = nrows = 10000
n_people = ncols = 23

birthdays = np.random.uniform(1,365,nrows*ncols).astype(int).reshape(nrows,ncols)
#birthdays